src/router.ts

```diff
@@ -0,0 +1,125 @@
+import Router from '@koa/router';
+import { z } from 'zod';
+import { fromZodError } from 'zod-validation-error';
+import zodToJsonSchema from 'zod-to-json-schema';
+import compose = require('koa-compose');
+import { IntrospectionConfig } from './koa';
+import { EndpointImplementation, implementRoute } from './koa-utils';
+import { IntrospectionResponse, OneSchemaDefinition } from './types';
+
+export type RouterEndpointDefinition<Name> = {
+  name: Name;
+  description?: string;
+  request: z.ZodType<any, any, any>;
+  response: z.ZodType<any, any, any>;
+};
+
+type Method = 'GET' | 'DELETE' | 'PUT' | 'POST' | 'PATCH';
+type RoughRoute = `${Method} ${string}`;
+
+type ZodSchema = {
+  [route: string]: RouterEndpointDefinition<string>;
+};
+
+export type OneSchemaRouterConfig<R extends Router<any, any>> = {
+  using: R;
+  introspection: IntrospectionConfig | undefined;
+};
+
+export class OneSchemaRouter<
+  Schema extends ZodSchema,
+  R extends Router<any, any>,
+> {
+  private router: R;
+
+  private constructor(
+    private schema: Schema,
+    { introspection, using: router }: OneSchemaRouterConfig<R>,
+  ) {
+    this.router = router;
+    if (introspection) {
+      router.get(introspection.route, (ctx, next) => {
+        const response: IntrospectionResponse = {
+          serviceVersion: introspection.serviceVersion,
+          schema: convertRouterSchemaToJSONSchemaStyle(this.schema),
+        };
+
+        ctx.body = response;
+        ctx.status = 200;
+        return next();
+      });
+    }
+  }
+
+  static create<R extends Router<any, any>>(
+    config: OneSchemaRouterConfig<R>,
+  ): OneSchemaRouter<{}, R> {
+    return new OneSchemaRouter({}, config);
+  }
```

Comment:
If we're using a class already, why not `create` when instantiating the class with `new`?

Reply:
Great question.

This `create` method allowed more predictably controlling the generic types of the created router. We _know_ that a new router has no routes/endpoints -- this signature reflects that (with the "hard-coded" `{}` on line 57).

If we just use a `constructor`, we can't "hard-code" that generic type into the router.